exonum/src/blockchain/mod.rs

```diff
@@ -226,24 +231,33 @@ impl BlockchainMut {
     }
 
     /// Creates and commits the genesis block with the given genesis configuration.
-    // TODO: extract genesis block into separate struct [ECR-3750]
-    fn create_genesis_block(
-        &mut self,
-        config: ConsensusConfig,
-        initial_services: Vec<InstanceConfig>,
-    ) -> Result<(), Error> {
-        config.validate()?;
+    fn create_genesis_block(&mut self, genesis_config: GenesisConfig) -> Result<(), Error> {
+        let GenesisConfig {
+            consensus_config,
+            artifacts,
+            builtin_instances,
+        } = genesis_config;
+        consensus_config.validate()?;
         let mut fork = self.fork();
-        Schema::new(&fork).consensus_config_entry().set(config);
+        Schema::new(&fork)
+            .consensus_config_entry()
+            .set(consensus_config);
+
+        for ArtifactSpec { artifact, payload } in artifacts {
+            if !self.dispatcher.is_artifact_deployed(&artifact) {
+                self.dispatcher
+                    .deploy_artifact_sync(&fork, artifact, payload)?;
+            }
+        }
 
         // Add service instances.
-        for instance_config in initial_services {
-            self.dispatcher.add_builtin_service(
-                &mut fork,
-                instance_config.instance_spec,
-                instance_config.artifact_spec.unwrap_or_default(),
-                instance_config.constructor,
-            )?;
+        for ConfiguredInstanceSpec {
```

Comment:
We don't really often (if do at all) use unpacking in exonum, so this code looks a bit inconsistent. Could it be avoided (here and in other places)?

Reply:
Well, perhaps I overused unpacking somewhere in code 🙂. But I'd like to keep it in loops because it saves us a temporary variable[s].